### PR TITLE
Conversation Sitching

### PR DESCRIPTION
**Pull request by codabrink**
https://github.com/xmtp/libxmtp/issues/1142

### CODE REVIEW

File: xmtp_mls/migrations/2024-12-18-170645_add_dm_id/up.sql [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+-- Fill the dm_id column
+UPDATE groups
+SET dm_id = 'dm:' ||
+    LOWER(
+        CASE
+            WHEN LOWER((SELECT inbox_id FROM identity)) < LOWER(dm_inbox_id)
+            THEN (SELECT inbox_id FROM identity) || ':' || dm_inbox_id
+            ELSE dm_inbox_id || ':' || (SELECT inbox_id FROM identity)
+        END
+    )
```
Comment (codabrink) [inline xmtp_mls/migrations/2024-12-18-170645_add_dm_id/up.sql:13]: Going to be writing a test to ensure this does what's expected before merging.

Comment (nplasterer) [reply xmtp_mls/migrations/2024-12-18-170645_add_dm_id/up.sql:13]: Do we still need this database migration? https://github.com/xmtp/libxmtp/pull/1374 Rich suggested we could just us the dm_inbox_id?

File: xmtp_mls/migrations/2024-12-18-170645_add_dm_id/up.sql [hunk 2]
```diff
@@ -0,0 +1,26 @@
+ALTER TABLE groups ADD COLUMN dm_id TEXT;
+ALTER TABLE groups ADD COLUMN last_message_ns BIGINT NOT NULL DEFAULT ((strftime('%s', 'now') * 1000000000) + (strftime('%f', 'now') * 1000000));
```
Comment (nplasterer) [inline xmtp_mls/migrations/2024-12-18-170645_add_dm_id/up.sql:2]: Is this going to conflict with the work @mchenani is doing for last message performance?

File: xmtp_mls/src/storage/encrypted_store/schema.rs [hunk 3]
```diff
@@ -52,9 +52,10 @@ diesel::table! {
         installations_last_checked -> BigInt,
         added_by_inbox_id -> Text,
         welcome_id -> Nullable<BigInt>,
-        dm_inbox_id -> Nullable<Text>,
```
Comment (nplasterer) [inline xmtp_mls/src/storage/encrypted_store/schema.rs:55]: I thought we exposed this out somewhere for easy look up of the the peerInboxId?

**Review by insipx — COMMENTED** [review]
Looks good so far, will give a deeper review tomorrow

File: xmtp_mls/migrations/2024-12-18-170645_add_dm_id/up.sql (hunk 1 above)
Comment (codabrink) [reply xmtp_mls/migrations/2024-12-18-170645_add_dm_id/up.sql:13]: Having a consistent key across DM's that should be stitched makes querying much simpler.

File: xmtp_mls/src/storage/encrypted_store/schema.rs (hunk 3 above)
Comment (nplasterer) [reply xmtp_mls/src/storage/encrypted_store/schema.rs:55]: All I really care about is this call still being performant when we list large lists of dms in RN where we have to serialize this field before sending. Can validate once we move this change up the chain.

**Comment by nplasterer** [comment]
Can you write up how push notifications will work for this?

If I have a list of conversations and the dms potentially have multiple topics how will I subscribe correctly to push notification topics?

File: xmtp_mls/migrations/2024-12-18-170645_add_dm_id/up.sql (hunk 2 above)
Comment (mchenani) [reply xmtp_mls/migrations/2024-12-18-170645_add_dm_id/up.sql:2]: Yes, it will!

File: xmtp_mls/migrations/2024-12-20-214747_add_dm_id/up.sql [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        gm.kind = 1
+)
+SELECT
+    g.id AS id,
+    g.created_at_ns,
+    g.membership_state,
+    g.installations_last_checked,
+    g.added_by_inbox_id,
+    g.welcome_id,
+    g.dm_id,
```
Comment (codabrink) [inline xmtp_mls/migrations/2024-12-20-214747_add_dm_id/up.sql:58]: Had to recreate the view to account for the column rename `dm_inbox_id` -> `dm_id`

**Review by nplasterer — APPROVED** [review]

File: xmtp_mls/src/storage/encrypted_store/mod.rs [hunk 5]
```diff
@@ -587,6 +591,83 @@ pub(crate) mod tests {
         EncryptedMessageStore::remove_db_files(db_path)
     }
 
+    #[tokio::test]
```
Comment (insipx) [inline xmtp_mls/src/storage/encrypted_store/mod.rs:594]: Test won't run in WASM without the `#[wasm_bindgen_test::wasm_bindgen_test(unsupported = tokio::test)]

**Review by insipx — APPROVED** [review]
Nice! Looks good other than nit on wasm test

**Pull request merged**